### PR TITLE
Enable Kaniko

### PR DESCRIPTION
**Pull request by vpnachev**
/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Enable Kaniko again.

We tried already with #85 but faced issues, therefore reverted to the previous build tool with #89.
@ccwienk took a look and identified that symbolic links are not handled very well with kaniko (they are pointing to a missing file).
Now, we introduce additional stage that retrieves properly the required files from the given `PROVIDER` folder.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```
/invite @ccwienk

### CODE REVIEW

File: .ci/pipeline_definitions [hunk 1]
```diff
@@ -73,7 +74,7 @@ terraformer:
           execute:
           - test-e2e.sh
           depends:
-          - publish
+          - build_oci_image_terraformer
```
Comment (ccwienk) [inline .ci/pipeline_definitions:77]: I suggest you rather change this to:
```
trait_depends:
- publish
```

Comment (ialidzhikov) [reply .ci/pipeline_definitions:77]: Does it work now? It was not working last time when I was checking it. And I asked you the very same question and back then you didn't replied anything. 😕

Comment (ccwienk) [reply .ci/pipeline_definitions:77]: @ialidzhikov : must have missed your comment back then. I just double-checked in our "cc-demo" repository

Comment (ialidzhikov) [reply .ci/pipeline_definitions:77]: Thank you for double-checking. It is great that we do not longer need this workaround. 👍

**Review by ccwienk — DISMISSED** [review]

File: Dockerfile [hunk 2]
```diff
@@ -3,7 +3,19 @@
...
+FROM golang-base AS terraform-config
+
+# overwrite to build provider-specific image
+ARG PROVIDER=all
+
+# Get entire dir to preserve the symblic links to TF_VERSION files
+COPY . .
+
+RUN cp ./build/$PROVIDER/TF_VERSION /TF_VERSION && \
+    cp ./build/$PROVIDER/terraform-bundle.hcl /terraform-bundle.hcl
```
Comment (ialidzhikov) [inline Dockerfile:18]: Thanks for the follow-up on this issue.
Another simpler option would be to remove the symlinks and each provider under `build/` to maintain its own TF_VERSION file? WDYT?

Comment (vpnachev) [reply Dockerfile:18]: Well, there are multiple options. I even have prepared another solution that was falling back to the `TF_VERSION` in the root of the repo. However, I like that with symbolic links we are maintaining the version just on one place, therefore I would like to preserve them.

**Comment by ccwienk** [comment]
@vpnachev @ialidzhikov : once you decided on which approach to take, would you also cherry-pick the resulting commit(s) to `refs/heads/release-v1.18`?

**Comment by mvladev** [comment]
One thing to be aware of - kaniko and containerd issue https://github.com/gardener/replica-reloader/issues/6

**Review by ialidzhikov — DISMISSED** [review]
/lgtm
/needs second-opinion
/invite @timebertt

**Comment by ccwienk** [comment]
> /lgtm
> /needs second-opinion
> /invite @timebertt

already fixed (https://github.com/gardener/cc-utils/commit/db40360dc9749674a6a935c84d7755ed32b63250)

File: .ci/pipeline_definitions [hunk 3]
```diff
@@ -58,7 +59,7 @@ terraformer:
               PROVIDER: slim
     steps:
       verify:
-        image: 'eu.gcr.io/gardener-project/3rd/golang:1.16.2'
+        image: 'golang:1.16.2'
```
Comment (timebertt) [inline .ci/pipeline_definitions:62]: Hmm, is the verify step also using a different runtime/config/environment that is not susceptible to docker hub rate limits? Why are we now able to switch back?

File: .test-defs/pod-e2e.yaml [hunk 4]
```diff
@@ -15,4 +15,4 @@ spec:
     ACCESS_KEY_ID_FILE=<(echo $ACCESS_KEY_ID)
     SECRET_ACCESS_KEY_FILE=<(echo $SECRET_ACCESS_KEY)
     IMAGE_TAG=$IMAGE_TAG
-  image: eu.gcr.io/gardener-project/3rd/golang:1.16.2
+  image: golang:1.16.2
```
Comment (timebertt) [inline .test-defs/pod-e2e.yaml:18]: Basically same here: The testmachinery cluster will probably also run into docker hub rate limits if we do it like this, right?

File: Dockerfile [hunk 5]
```diff
@@ -3,7 +3,19 @@
...
+FROM golang:1.16.2 AS golang-base
+
+############# terraform-config
+FROM golang-base AS terraform-config
+
+# overwrite to build provider-specific image
+ARG PROVIDER=all
+
+# Get entire dir to preserve the symblic links to TF_VERSION files
+COPY . .
```
Comment (timebertt) [inline Dockerfile:15]: Can we avoid copying the entire directory (including `vendor` and so on) and instead only copy the `build` dir / relevant files?

File: .ci/pipeline_definitions (hunk 3 above)
Comment (ccwienk) [reply .ci/pipeline_definitions:62]: @timebertt : I think for the build-steps, this should not be such a big issue, as concourse does cache the images.

the actual quota-issues stemmed from the image-builds. With the switch to kaniko-builder, however, we have the possibility to authenticate against all required source registries, so this issue should be a relict from the past :-)

File: .test-defs/pod-e2e.yaml (hunk 4 above)
Comment (ccwienk) [reply .test-defs/pod-e2e.yaml:18]: see above

**Comment by ccwienk** [comment]
@vpnachev @timebertt @ialidzhikov : any progress w.r.t. the open discussion? This is the last repository (I currently know of) that blocks switching to kaniko-builder as default :-/

**Comment by timebertt** [comment]
> @vpnachev @timebertt @ialidzhikov : any progress w.r.t. the open discussion? This is the last repository (I currently know of) that blocks switching to kaniko-builder as default :-/

Are we already using kaniko for g/g?

**Comment by ccwienk** [comment]
> Are we already using kaniko for g/g?

@timebertt : no. However g/g is already prepared for kaniko-builds (as in: will not break due to dependency against `publish` step)

File: Dockerfile (hunk 5 above)
Comment (vpnachev) [reply Dockerfile]: Done.

File: Dockerfile (hunk 2 above)
Comment (vpnachev) [reply Dockerfile]: I have removed the symlinks and each provider has its own regular TF_VERSION file.

Comment (ialidzhikov) [reply Dockerfile]: Thanks! I assume we can also remove the TF_VERSION file in the repo root (and the reference in `.dockerignore`)?

**Review by ccwienk — DISMISSED** [review]

File: Dockerfile (hunk 2 above)
Comment (vpnachev) [reply Dockerfile]: Thanks! Also done.

**Review by ialidzhikov — APPROVED** [review]
/lgtm
/squash

**Pull request merged**